### PR TITLE
feature/notifications-friend-requests

### PR DESCRIPTION
**Pull request by matt-opw**
# Add Friend Request Push Notifications (Backend + Android Integration)

## What  
This PR implements the **friend-request push notifications**, including:

### Backend (Python Firebase Functions)
- New callable Function `send_friend_request_notification`).  

### Android
- New `FirebaseFriendRequestNotifier` that triggers the Cloud Function.  
- MainActivity now supports **navigation triggered by notification payloads**.  

---

## Why  
Previously, friend requests produced no visible notifications for the receiving user, the receiver was notify only via the pop up if he was in the app.  
This PR adds:

- real push notifications when a friend request is sent  
- automatic navigation when the app is opened from the notification  

---

## How  
- Added the Python backend Cloud Function  
- Added the Android notifier to call it  
- Updated MainActivity to parse and react to notification intents  

---

## Tests  
In addition to test files some manual testing have been performed

---

## Screenshot  
Capture d’écran 2025-12-10 à 17 40 39

---

## About the Low Line Coverage (~56%)  
The current coverage drop is **due to**:

### **1. Python Firebase Functions are not counted**  
All new backend logic is in Python, so JaCoCo cannot instrument or measure it.

### **2. Tests annotated with `@RequiresFirebaseFunctions` were removed from the CI**  
These tests depended on executing Python Cloud Functions, but the CI emulator only supports Node.js.  
They were removed to stabilize CI, which naturally reduces the number of covered lines.  

---

## About `@RequiresFirebaseFunctions`  
This annotation marked tests that required Firebase Functions to run.  
Since Python Functions cannot be executed in CI, tests using this annotation must be skipped or removed, meaning their corresponding code paths remain **uncovered**. However these tests should still **pass in local computers** !

---

[PR description generated with help of AI and the log file]

### CODE REVIEW

File: app/src/main/java/com/android/mygarden/model/notifications/FirebaseFriendRequestNotifier.kt [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package com.android.mygarden.model.notifications
+
+import com.google.firebase.functions.FirebaseFunctions
+import kotlinx.coroutines.tasks.await
+
+/** Notifier implemented by Firebase functions. */
```
Comment (yngflibustier) [inline app/src/main/java/com/android/mygarden/model/notifications/FirebaseFriendRequestNotifier.kt]: As the implementation really does a Firebase call I would suggest to enhance the KDoc to indicate to users that this actual implementation can throw exceptions so they know they should wrap it inside of a try/catch block.

File: app/src/main/java/com/android/mygarden/model/notifications/FirebaseFriendRequestNotifier.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+class FirebaseFriendRequestNotifier : FriendRequestNotifier {
+  override suspend fun notifyRequestSent(targetUid: String, fromPseudo: String) {
+    FirebaseFunctions.getInstance()
+        .getHttpsCallable("send_friend_request_notification")
```
Comment (yngflibustier) [inline app/src/main/java/com/android/mygarden/model/notifications/FirebaseFriendRequestNotifier.kt]: The name of the function you call in `getHttpsCallable` is hardcoded, I suggest to put it in a companion object to avoid this "magic name" (to respect best coding principles).

File: functions/main.py [hunk 3]
```diff
@@ -8,14 +8,14 @@
...
 from firebase_functions import scheduler_fn
+from firebase_functions import https_fn, options
 from firebase_admin import initialize_app, firestore, messaging
 from datetime import datetime, timezone, timedelta
 from google.cloud.firestore_v1.base_document import DocumentSnapshot
 import functools, random, logging, time
 
-
+initialize_app()
 # Initialize these globally, but with None, so they don't run on import
```
Comment (yngflibustier) [inline functions/main.py]: Change comment stating that the app is not run on import (with new modifications only db don't run on import).

File: app/src/androidTest/java/com/android/mygarden/model/notifications/FriendRequestsFirebaseNotifierTest.kt
Comment (yngflibustier) [inline app/src/androidTest/java/com/android/mygarden/model/notifications/FriendRequestsFirebaseNotifierTest.kt]: LGTM.

File: app/src/androidTest/java/com/android/mygarden/ui/addFriend/AddFriendVMonAskTest.kt [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        onError = {
+          error = true
+          latch.countDown()
+        })
+
+    assertTrue(latch.await(5, TimeUnit.SECONDS))
+    assertFalse(error)
+    assertTrue(success)
+
+    println("DEBUG fakeRequests size = ${fakeRequests.incomingRequestsFlow.value.size}")
```
Comment (yngflibustier) [inline app/src/androidTest/java/com/android/mygarden/ui/addFriend/AddFriendVMonAskTest.kt]: Are those DEBUG print still needed for the production test code ?

File: app/src/androidTest/java/com/android/mygarden/ui/addFriend/AddFriendVMonAskTest.kt
Comment (yngflibustier) [inline app/src/androidTest/java/com/android/mygarden/ui/addFriend/AddFriendVMonAskTest.kt]: Tests for the client-side requests seem good and you have tried many edge cases !

File: functions/main.py [hunk 5]
```diff
@@ -8,14 +8,14 @@
 
 from enum import Enum
 from firebase_functions import scheduler_fn
+from firebase_functions import https_fn, options
```
[the remark this answers is not in the document]
Comment (matt-opw) [reply functions/main.py]: Yep right ! I didn't use it.

File: functions/main.py [hunk 6]
```diff
@@ -93,13 +91,97 @@ def _get_user_token(uid: str) -> str | None:
...
+        return False
+
+    message = messaging.Message(
+        token=token,
+        notification=messaging.Notification(
+            title="New Friend Request 🤝",
+            body=f"{from_pseudo} wants to be your friend!"
+        ),
+        data={
+                "type": "FRIEND_REQUEST",
```
[the remark this answers is not in the document]
Comment (matt-opw) [reply functions/main.py]: Not done in purpose ... Thanks !

File: functions/main.py (hunk 3 above)
Comment (matt-opw) [reply functions/main.py]: Done !

File: app/src/main/java/com/android/mygarden/ui/addFriend/AddFriendScreen.kt
[the remark this answers is not in the document]
Comment (matt-opw) [reply app/src/main/java/com/android/mygarden/ui/addFriend/AddFriendScreen.kt]: Right !

File: app/src/main/java/com/android/mygarden/ui/addFriend/AddFriendViewModel.kt
[the remark this answers is not in the document]
Comment (matt-opw) [reply app/src/main/java/com/android/mygarden/ui/addFriend/AddFriendViewModel.kt]: Done !

File: app/src/androidTest/java/com/android/mygarden/ui/addFriend/AddFriendVMonAskTest.kt [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+
+    assertEquals(1, fakeRequests.incomingRequestsFlow.value.size)
+    println("DEBUG fakeRequests value = ${fakeRequests.incomingRequestsFlow.value[0].fromUserId}")
+
+    assertEquals("target", fakeRequests.incomingRequestsFlow.value[0].fromUserId)
```
[the remark this answers is not in the document]
Comment (matt-opw) [reply app/src/androidTest/java/com/android/mygarden/ui/addFriend/AddFriendVMonAskTest.kt]: I didn't even made this test. I just move it from test to androidTest since I use emulators. You are right, the fake FakeFriendRequestsRepository is wrong. This should be fixed in next week's sprint.

File: app/src/test/java/com/android/mygarden/utils/HelperForAddFriendTests.kt
[the remark this answers is not in the document]
Comment (matt-opw) [reply app/src/test/java/com/android/mygarden/utils/HelperForAddFriendTests.kt]: I only move it here since I needed some of the functions, but you are right. However I think it is good to keep it so that it is consistent with HelperForAddFriendTests in the test.

File: app/src/main/java/com/android/mygarden/model/notifications/FirebaseFriendRequestNotifier.kt (hunk 1 above)
Comment (matt-opw) [reply app/src/main/java/com/android/mygarden/model/notifications/FirebaseFriendRequestNotifier.kt]: Done !

File: app/src/main/java/com/android/mygarden/model/notifications/FirebaseFriendRequestNotifier.kt (hunk 2 above)
Comment (yngflibustier) [reply app/src/main/java/com/android/mygarden/model/notifications/FirebaseFriendRequestNotifier.kt]: Okay that is not a big deal !

File: app/src/androidTest/java/com/android/mygarden/ui/addFriend/AddFriendVMonAskTest.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (yngflibustier) [reply app/src/androidTest/java/com/android/mygarden/ui/addFriend/AddFriendVMonAskTest.kt]: Okay yes not a big deal as long as it is fixed ! Yes, your are right I just saw that the Fake repository is a bit suspicious...
